### PR TITLE
[feature] Drag & Drop designer form actions

### PR DESCRIPTION
**Pull request by elpaso**
## Drag & Drop designer form actions

form_actions

### CODE REVIEW

**Comment by nirvn** [comment]
@elpaso , cool idea!

**Comment by elpaso** [comment]
> @elpaso , cool idea!

Not mine's idea but yeah, I like it too. 

It was actually @andreasneumann 's idea

**Comment by nyalldawson** [comment]
Nice! Are these hidden in search/multiedit/edit modes?

File: src/gui/qgsattributeform.cpp [hunk 1]
```diff
@@ -1346,8 +1348,8 @@ void QgsAttributeForm::synchronizeState()
 
   for ( QgsWidgetWrapper *ww : std::as_const( mWidgets ) )
   {
-    QgsEditorWidgetWrapper *eww = qobject_cast<QgsEditorWidgetWrapper *>( ww );
-    if ( eww )
+
+    if ( QgsEditorWidgetWrapper *eww = qobject_cast<QgsEditorWidgetWrapper *>( ww ); eww )
```
Comment (3nids) [inline src/gui/qgsattributeform.cpp:1352]: why this change? I find this less readable.

Comment (nyalldawson) [reply src/gui/qgsattributeform.cpp:1352]: It doesn't need the "; eww" part, it works without that

**Comment by elpaso** [comment]
> Nice! Are these hidden in search/multiedit/edit modes?

The widget is only visible in SingleEditMode and AddFeatureMode.

File: src/gui/vector/qgsattributesformproperties.cpp [hunk 2]
```diff
@@ -662,6 +705,13 @@ QgsAttributeEditorElement *QgsAttributesFormProperties::createAttributeEditorWid
       break;
     }
 
+    case DnDTreeItemData::Action:
+    {
+      const QgsAction action { mLayer->actions()->action( itemData.name() )};
```
Comment (3nids) [inline src/gui/vector/qgsattributesformproperties.cpp:710]:
```suggestion
      const QgsAction action = mLayer->actions()->action( itemData.name() );
```
I don't have a strong technical background but I'd like our code to be consistent and maybe discuss before starting integrating these init methods which I find less readable.

File: src/gui/vector/qgsattributesformproperties.cpp [hunk 3]
```diff
@@ -548,6 +586,11 @@ void QgsAttributesFormProperties::loadAttributeSpecificEditor( QgsAttributesDnDT
         loadAttributeContainerEdit();
         break;
       }
+      case DnDTreeItemData::Action:
+      {
+        receiver->clearSelection();
```
Comment (3nids) [inline src/gui/vector/qgsattributesformproperties.cpp:591]: why not calling selectFirstMatchingItem here?

File: src/gui/vector/qgsattributesformproperties.cpp [hunk 4]
```diff
@@ -410,6 +431,23 @@ QTreeWidgetItem *QgsAttributesFormProperties::loadAttributeEditorTreeItem( QgsAt
       break;
     }
 
+    case QgsAttributeEditorElement::AeTypeAction:
+    {
+      const QgsAttributeEditorAction *actionEditor = static_cast<const QgsAttributeEditorAction *>( widgetDef );
+      const QgsAction action { actionEditor->action( mLayer ) };
```
Comment (3nids) [inline src/gui/vector/qgsattributesformproperties.cpp:437]:
```suggestion
      const QgsAction action = actionEditor->action( mLayer );
```

File: src/gui/vector/qgsattributeactiondialog.cpp [hunk 5]
```diff
@@ -224,7 +225,9 @@ void QgsAttributeActionDialog::swapRows( int row1, int row2 )
 
 QgsAction QgsAttributeActionDialog::rowToAction( int row ) const
 {
-  QgsAction action( static_cast<QgsAction::ActionType>( mAttributeActionTable->item( row, Type )->data( Qt::UserRole ).toInt() ),
+  const QUuid id { mAttributeActionTable->item( row, Type )->data( Qt::UserRole + 1 ).toUuid() };
```
Comment (3nids) [inline src/gui/vector/qgsattributeactiondialog.cpp:228]:
```suggestion
  const QUuid id = mAttributeActionTable->item( row, Type )->data( Qt::UserRole + 1 ).toUuid();
```

File: src/gui/vector/qgsattributeactiondialog.cpp [hunk 6]
```diff
@@ -112,6 +112,7 @@ void QgsAttributeActionDialog::insertRow( int row, const QgsAction &action )
   // Type
   item = new QTableWidgetItem( textForType( action.type() ) );
   item->setData( Qt::UserRole, action.type() );
+  item->setData( Qt::UserRole + 1, action.id() );
```
Comment (3nids) [inline src/gui/vector/qgsattributeactiondialog.cpp:115]: could you define a private enum for these two (ActionType, ActionId) ?

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    else
+    {
+      mActionButton->setText( shortTitle.isEmpty() ? description : shortTitle );
+      if ( ! shortTitle.isEmpty() )
+      {
+        mActionButton->setToolTip( description );
+      }
+    }
+
+    if ( mAction.isEnabledOnlyWhenEditable() && ! layer()->isEditable() )
```
Comment (3nids) [inline src/gui/editorwidgets/qgsactionwidgetwrapper.cpp:94]:
```suggestion
    if ( mAction.isEnabledOnlyWhenEditable() && !layer()->isEditable() )
```

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 8]
```diff
@@ -444,7 +444,7 @@ void QgsValueRelationWidgetWrapper::showIndeterminateState()
     {
       for ( int i = 0; i < nofColumns; ++i )
       {
-        whileBlocking( mTableWidget )->item( j, i )->setCheckState( Qt::PartiallyChecked );
+        whileBlocking( mTableWidget ) ->item( j, i )->setCheckState( Qt::PartiallyChecked );
```
Comment (3nids) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp:447]:
```suggestion
        whileBlocking( mTableWidget )->item( j, i )->setCheckState( Qt::PartiallyChecked );
```

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp (hunk 7 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsactionwidgetwrapper.cpp:94]: also, following the logic further down, you should also check for the form mode?
(see https://github.com/qgis/QGIS/pull/44710/files#diff-63b2666af07d22ef212357c4b9c789a154f4c54a35429d09f0800471c1304bafR1345)

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  mActionButton = qobject_cast<QPushButton *>( editor );
+
+  if ( !mActionButton )
+    return;
+
+  if ( valid() && layer() )
+  {
+    const QString shortTitle { mAction.shortTitle() }; // might be empty
+    const QString description { mAction.name() };  // mandatory
+    const QIcon icon { mAction.icon() };  // might be invalid
```
Comment (3nids) [inline src/gui/editorwidgets/qgsactionwidgetwrapper.cpp:77]:
```suggestion
    const QString shortTitle = mAction.shortTitle(); // might be empty
    const QString description = mAction.name();  // mandatory
    const QIcon icon = mAction.icon();  // might be invalid
```

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    // Configure push button
+    if ( ! icon.isNull() )
+    {
+      mActionButton->setIcon( mAction.icon() );
```
Comment (3nids) [inline src/gui/editorwidgets/qgsactionwidgetwrapper.cpp:82]:
```suggestion
      mActionButton->setIcon( icon );
```
?

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+void QgsActionWidgetWrapper::setFeature( const QgsFeature &feature )
+{
+  mFeature = feature;
+}
+
+void QgsActionWidgetWrapper::setEnabled( bool enabled )
+{
+  if ( valid() && layer() )
+  {
+    mActionButton->setEnabled( ! mAction.isEnabledOnlyWhenEditable() || enabled );
```
Comment (3nids) [inline src/gui/editorwidgets/qgsactionwidgetwrapper.cpp:51]:
```suggestion
    mActionButton->setEnabled( !mAction.isEnabledOnlyWhenEditable() || enabled );
```

File: src/core/qgsaction.h [hunk 12]
```diff
@@ -119,6 +119,48 @@ class CORE_EXPORT QgsAction
...
+      , mShortTitle( shortTitle )
+      , mIcon( icon )
+      , mCommand( action )
+      , mCaptureOutput( capture )
+      , mActionScopes( actionScopes )
+      , mNotificationMessage( notificationMessage )
+      , mId( id )
+      , mIsEnabledOnlyWhenEditable( enabledOnlyWhenEditable )
+    {}
+#endif
```
Comment (3nids) [inline src/core/qgsaction.h:162]:
```suggestion
    QgsAction( const QUuid &id, ActionType type, const QString &description, const QString &action, const QString &icon, bool capture, const QString &shortTitle = QString(), const QSet<QString> &actionScopes = QSet<QString>(), const QString &notificationMessage = QString(), bool enabledOnlyWhenEditable SIP_PYARGREMOVE = false )
      : mType( type )
      , mDescription( description )
      , mShortTitle( shortTitle )
      , mIcon( icon )
      , mCommand( action )
      , mCaptureOutput( capture )
      , mActionScopes( actionScopes )
      , mNotificationMessage( notificationMessage )
      , mId( id )
      , mIsEnabledOnlyWhenEditable( enabledOnlyWhenEditable )
    {}
```

can you try this with SIP_PYARGREMOVE?
if this doesn't work, you can at least remove the init code from the SIP_RUN part as it is useless (see that you left a non-inited variable `enabledOnlyWhenEditable`)

you can apply the same correction to the other constructor.

**Review by 3nids — COMMENTED** [review]
mainly minor code formatting comments.

What is you reasoning for using {}-initialisation? I am not sure if we discussed this in the past. I find this less readable and it is not coherent with the rest of the code base. 
If there is a good technical reason, then let's discuss/go for it. Otherwise, I'd prefer to stick to the current approach. But it is not my decision to take, just my 2 cents.

File: src/gui/vector/qgsattributesformproperties.cpp (hunk 2 above)
Comment (elpaso) [reply src/gui/vector/qgsattributesformproperties.cpp:710]: > mainly minor code formatting comments.
> 
> What is you reasoning for using {}-initialisation? I am not sure if we discussed this in the past. I find this less readable and it is not coherent with the rest of the code base.
> If there is a good technical reason, then let's discuss/go for it. Otherwise, I'd prefer to stick to the current approach. But it is not my decision to take, just my 2 cents.

@3nids there have been some discussion about uniform initialization (aka brace-initialization), I believe we agreed upon not using it when initializing containers (because it may lead to unexpected results on some compilers), for the rest it's free to use. 

There are pros and cons by using braces, I personally like the fact that they prohibit narrowing and avoid the "most vexing parse" issue, this has helped me to find a few bugs with loss of precision.

Comment (elpaso) [reply src/gui/vector/qgsattributesformproperties.cpp:710]: BTW it's a little late because brace initilizers are used everywhere in QGIS code.

File: src/core/qgsaction.h (hunk 12 above)
Comment (elpaso) [reply src/core/qgsaction.h:162]: > can you try this with SIP_PYARGREMOVE?
> if this doesn't work, you can at least remove the init code from the SIP_RUN part as it is useless (see that you left a non-inited variable `enabledOnlyWhenEditable`)
> 
> you can apply the same correction to the other constructor.

Yeah, I just copied the already existing ctors, I also wondered why they were like that.

File: src/gui/vector/qgsattributesformproperties.cpp (hunk 3 above)
Comment (elpaso) [reply src/gui/vector/qgsattributesformproperties.cpp]: done

File: src/gui/vector/qgsattributeactiondialog.cpp (hunk 6 above)
Comment (elpaso) [reply src/gui/vector/qgsattributeactiondialog.cpp]: done

File: src/gui/editorwidgets/qgsactionwidgetwrapper.cpp (hunk 7 above)
Comment (elpaso) [reply src/gui/editorwidgets/qgsactionwidgetwrapper.cpp]: I originally did that, but I changed it when I realized (if I'm not wrong) that the mode can change without `initWidget` being called again, the lambda in the ctor takes care of updating the widget status (visibility) dynamically whenever the form mode changes. The editable state is kept in sync by `QgsActionWidgetWrapper::setEnabled`

**Comment by elpaso** [comment]
@3nids thank you for the review, I think I have addressed all your comments except for the brace initializers.

**Comment by nyalldawson** [comment]
> What is you reasoning for using {}-initialisation? I am not sure if we discussed this in the past. I find this less readable and it is not coherent with the rest of the code base.

I personally don't have any issue with using this in certain circumstances -- eg they are very useful for list/map creation! My concerns are using it with QVariant and the json classes (which have unpredictable behaviour), and the fragile nature of how astyle formats the resultant code (which means that locally parts of files get changed by the pre commit script, which CI then rejects). 

We're stuck in an awkward situation -- we can't move wholesale to brace initialisation due to the qvariant/json blockers, so we'll either need to decide to live with mixed use or come up with a proper policy of when they are acceptable...

**Comment by elpaso** [comment]
> > What is you reasoning for using {}-initialisation? I am not sure if we discussed this in the past. I find this less readable and it is not coherent with the rest of the code base.
> 
> I personally don't have any issue with using this in certain circumstances -- eg they are very useful for list/map creation! My concerns are using it with QVariant and the json classes (which have unpredictable behaviour), and the fragile nature of how astyle formats the resultant code (which means that locally parts of files get changed by the pre commit script, which CI then rejects).
> 
> We're stuck in an awkward situation -- we can't move wholesale to brace initialisation due to the qvariant/json blockers, so we'll either need to decide to live with mixed use or come up with a proper policy of when they are acceptable...

My personal criteria is to use braces everywhere except for:
- containers (because they may have an overloaded ctor that interpret the second argument as a size of the container)
- QVariants (because they may implicitly create a QVariantList)

We already have a test for the latter: https://github.com/qgis/QGIS/blob/master/tests/code_layout/test_qvariant_no_brace_init.sh

The issue with containers is actually a very rare corner case, but to stay on the safe side it's better to not use braces.

**Comment by m-kuhn** [comment]
Thanks for this pull request @elpaso , really nice functionality.
Thanks for the review @3nids 

Regarding the `{}` questions.

To me

> they are very useful for list/map creation

and 

> containers (because they may have an overloaded ctor that interpret the second argument as a size of the container)

seem to be contradictory.
Am I missing something here? Or are they really complex to understand and get right?

**Comment by 3nids** [comment]
To me, here would be a short summary regarding braces initialisations:
* they are less readable than usual init
* they are useful for lists/maps
* they might prevent errors in edge cases (narrowing, most-vexing parse).
* they might produces errors in edge cases (containers?, QVariant, json)

Some curly init were also apparently failing with Qt6: https://github.com/qgis/QGIS/pull/44205/files

I would propose to allow them for the cases where they are useful AND safe only.

But this is not my decision to take and I don't feel I have the knowledge for a statement at the moment.
We need to define the coding guidelines here.
I guess that would be up to the PSC to decide and/or form a group of expert to formulate a decision.
I will forward this discussion to the PSC and ask for feedback.

In the mean time, can we hold the introduction of these?
As far as I see in this PR, there is no risk of narrowing or most-vexing parse errors.

**Comment by elpaso** [comment]
> Thanks for this pull request @elpaso , really nice functionality.
> Thanks for the review @3nids
> 
> Regarding the `{}` questions.
> 
> To me
> 
> > they are very useful for list/map creation
> 
> and
> 
> > containers (because they may have an overloaded ctor that interpret the second argument as a size of the container)
> 
> seem to be contradictory.
> Am I missing something here? Or are they really complex to understand and get right?

No, I didn't write the right thing:  what I wanted to say is that if a container has an overloaded ctor the two forms have a complete different meaning.

```c++
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    std::vector<int> v{10};
    cout << "v size: "  << v.size() << std::endl;

    std::vector<int> v2(10);
    cout << "v2 size: "  << v2.size() << std::endl;

    return 0;
}

// Output:
v size: 1
v2 size: 10
```

**Comment by elpaso** [comment]
> To me, here would be a short summary regarding braces initialisations:
> 
>     * they are less readable than usual init
> 
>     * they are useful for lists/maps
> 
>     * they might prevent errors in edge cases (narrowing, most-vexing parse).
> 
>     * they might produces errors in edge cases (containers?, QVariant, json)
> 
> 
> Some curly init were also apparently failing with Qt6: https://github.com/qgis/QGIS/pull/44205/files
> 
> I would propose to allow them for the cases where they are useful AND safe only.
> 
> But this is not my decision to take and I don't feel I have the knowledge for a statement at the moment.
> We need to define the coding guidelines here.
> I guess that would be up to the PSC to decide and/or form a group of expert to formulate a decision.
> I will forward this discussion to the PSC and ask for feedback.
> 
> In the mean time, can we hold the introduction of these?
> As far as I see in this PR, there is no risk of narrowing or most-vexing parse errors.

As I have already mentioned, it is a little late for this: brace initialization has been used for years through al QGIS for years now.

Besides that we have a nice mixture or `=` and `()` initialization too:

```c++
// all the same:
int i = 0;
int i(0);
int i{0};
```

I say everyone uses what he likes most, just avoid the pitfalls of QVariant(s) and pay attention to containers.

In any event can we please move forward with this PR and open a discussion about brace initialization in another context?

**Comment by 3nids** [comment]
Yes, discussion moved to the PSC mailing list.

**Review by m-kuhn — APPROVED** [review]
Let's move on with this, all critical review comments have been solved and the discussion about brace initialization is unrelated.
Thanks again, this feature is really neat
